src/proxy/inbound.rs

```diff
@@ -78,67 +82,58 @@ impl Inbound {
-            debug!(%conn, "accepted connection");
-            let enable_original_source = self.cfg.enable_original_source;
-            let metrics = self.metrics.clone();
-            async move {
-                Ok::<_, hyper::Error>(service_fn(move |req| {
-                    Self::serve_connect(
-                        workloads.clone(),
-                        conn.clone(),
-                        enable_original_source.unwrap_or_default(),
-                        req,
-                        metrics.clone(),
-                    )
-                }))
-            }
-        });
-
+        // let (tx, rx) = oneshot::channel();
         let acceptor = InboundCertProvider {
             workloads: self.workloads.clone(),
             cert_manager: self.cert_manager.clone(),
         };
-        let tls_stream = crate::hyper_util::tls_server(acceptor, self.listener);
-        let incoming = hyper::server::accept::from_stream(tls_stream);
-
-        let server = Server::builder(incoming)
-            .http2_only(true)
-            .http2_initial_stream_window_size(self.cfg.window_size)
-            .http2_initial_connection_window_size(self.cfg.connection_window_size)
-            .http2_max_frame_size(self.cfg.frame_size)
-            .serve(service)
-            .with_graceful_shutdown(async {
-                // Wait until the drain is signaled
-                let shutdown = self.drain.signaled().await;
-                // Once `shutdown` is dropped, we are declaring the drain is complete. Hyper will start draining
-                // once with_graceful_shutdown function exists, so we need to exit the function but later
-                // drop `shutdown`.
-                if tx.send(shutdown).is_err() {
-                    error!("HBONE receiver dropped")
+        let workloads = self.workloads;
+        let drain_stream = self.drain.clone();
+        let stream = crate::hyper_util::tls_server(acceptor, self.listener);
+        let mut stream = stream.take_until(Box::pin(drain_stream.signaled()));
+        while let Some(socket) = stream.next().await {
+            let workloads = workloads.clone();
+            let metrics = self.metrics.clone();
+            let drain = self.drain.clone();
+            tokio::task::spawn(async move {
+                let dst = crate::socket::orig_dst_addr_or_default(socket.get_ref());
+                let conn = rbac::Connection {
+                    src_identity: socket
+                        .ssl()
+                        .peer_certificate()
+                        .and_then(|x| crate::tls::boring::extract_sans(&x).first().cloned()),
+                    src_ip: to_canonical(socket.get_ref().peer_addr().unwrap()).ip(),
+                    dst,
+                };
+                debug!(%conn, "accepted connection");
+                let enable_original_source = self.cfg.enable_original_source;
+                let server = crate::hyper_util::http2_server()
```

Comment:
same here.. it's a "serverside connection" while this whole piece of machinery to accept connections feels like the server.